text/0000-inline-asm.md

```diff
@@ -0,0 +1,869 @@
+
+```rust
+fn mul(a: u32, b: u32) -> u64 {
+    let (lo, hi) = unsafe {
+        asm!("mul {}", in(reg) a, in("eax") b, lateout("eax"), lateout("edx"))
+    };
+
+    hi as u64 << 32 + lo as u64
+}
+```
+
+# Prior art
+[prior-art]: #prior-art
+
+## GCC inline assembly
+
+The proposed syntax is very similar to GCC's inline assembly in that it is based on string substitution while leaving actual interpretation of the final string to the assembler. However GCC uses poorly documented single-letter constraint codes and template modifiers. Clang tries to emulate GCC's behavior, but there are still several cases where its behavior differs from GCC's.
+
+The main reason why this is so complicated is that GCC's inline assembly basically exports the raw internals of GCC's register allocator. This has resulted in many internal constraint codes and modifiers being widely used, despite them being completely undocumented.
+
+## D & MSVC inline assembly
+
+See the section [above][dsl].
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+None
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Flag outputs
+
+GCC supports a special type of output which allows an asm block to return a `bool` encoded in the condition flags register. This allows the compiler to branch directly on the condition flag instead of materializing the condition as a `bool`.
+
+We can support this in the future with a special output operand type.
+
+## `asm goto`
+
+GCC supports passing C labels (the ones used with `goto`) to an inline asm block, with an indication that the asm code may jump directly to one of these labels instead of leaving the asm block normally.
+
+This could be supported by allowing code blocks to be specified as operand types. The following code will print `a` if the input value is `42`, or print `b` otherwise.
+
+```rust
+asm!("cmp {}, 42; jeq {}",
+    in(reg) val,
+    label { println!("a"); },
+    fallthrough { println!("b"); }
+);
+```
+
+## Unique ID per `asm`
+
+GCC supports `%=` which generates a unique identifier per instance of an asm block. This is guaranteed to be unique even if the asm block is duplicated (e.g. because of inlining).
+
+We can support this in the future with a special operand type.
+
+## `imm` and `sym` for `global_asm!`
+
+The `global_asm!` macro could be extended to support `imm` and `sym` operands since those can be resolved by simple string substitution. Symbols used in `global_asm!` will be marked as `#[used]` to ensure that they are not optimized away by the compiler.
+
+## Memory operands
+
+We could support `mem` as an alternative to specifying a register class which would leave the operand in memory and instead produce a memory address when inserted into the asm string. This would allow generating more efficient code by taking advantage of addressing modes instead of using an intermediate register to hold the computed address.
```

Comment:
It's very common to have instructions (like `add` or `mov`) that accept inputs as both registers or memory. Specifying this as a `rm` constraint allows the compiler to do better register allocation/spilling.

That being said, LLVM currently ignores any freedom in the constraints and always picks memory:

> Thus, it simply tries to make a choice that’s most likely to compile, not one that will be optimal performance. (e.g., given “rm”, it’ll always choose to use memory, not registers).
> - [Section 'Constraint codes'](https://releases.llvm.org/9.0.0/docs/LangRef.html#module-level-inline-assembly)

Reply:
Also LLVM only really supports memory addressing modes for inline asm on x86. On ARM/AArch64 it will just perform the address calculation and put the result in a register.